processor/internal/transformer/sourcehydration/source_hydration.go

```diff
@@ -0,0 +1,288 @@
+package sourcehydration
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	obskit "github.com/rudderlabs/rudder-observability-kit/go/labels"
+
+	"github.com/cenkalti/backoff"
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/jsonrs"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+
+	transformerclient "github.com/rudderlabs/rudder-server/internal/transformer-client"
+	transformerutils "github.com/rudderlabs/rudder-server/processor/internal/transformer"
+	"github.com/rudderlabs/rudder-server/processor/types"
+	"github.com/rudderlabs/rudder-server/utils/httputil"
+)
+
+const srcHydrationStage = "source_hydration"
+
+type Opt func(*Client)
+
+func WithClient(client transformerclient.Client) Opt {
+	return func(s *Client) {
+		s.client = client
+	}
+}
+
+// Client handles source hydration transformations
+type Client struct {
+	config struct {
+		sourceHydrationURL      string
+		maxRetry                config.ValueLoader[int]
+		failOnError             config.ValueLoader[bool]
+		maxRetryBackoffInterval config.ValueLoader[time.Duration]
+		timeoutDuration         time.Duration
+		batchSize               config.ValueLoader[int]
+	}
+	conf   *config.Config
+	log    logger.Logger
+	stat   stats.Stats
+	client transformerclient.Client
+}
+
+// New creates a new source hydration client
+func New(conf *config.Config, log logger.Logger, stat stats.Stats, opts ...Opt) *Client {
+	handle := &Client{}
+	handle.conf = conf
+	handle.log = log
+	handle.stat = stat
+	handle.client = transformerclient.NewClient(transformerutils.TransformerClientConfig(conf, "SourceHydration"))
+	handle.config.sourceHydrationURL = handle.conf.GetString("DEST_TRANSFORM_URL", "http://localhost:9090")
+	handle.config.timeoutDuration = conf.GetDuration("HttpClient.procTransformer.timeout", 600, time.Second)
+	handle.config.maxRetry = conf.GetReloadableIntVar(30, 1, "Processor.SourceHydration.maxRetry", "Processor.maxRetry")
+	handle.config.failOnError = conf.GetReloadableBoolVar(false, "Processor.SourceHydration.failOnError", "Processor.Transformer.failOnError")
+	handle.config.maxRetryBackoffInterval = conf.GetReloadableDurationVar(30, time.Second, "Processor.SourceHydration.maxRetryBackoffInterval", "Processor.maxRetryBackoffInterval")
+	handle.config.batchSize = conf.GetReloadableIntVar(100, 1, "Processor.SourceHydration.batchSize", "Processor.transformBatchSize")
+
+	for _, opt := range opts {
+		opt(handle)
+	}
+
+	return handle
+}
+
+// Hydrate performs source hydration on the provided events
+func (c *Client) Hydrate(ctx context.Context, hydrationReq types.SrcHydrationRequest) (types.SrcHydrationResponse, error) {
```

Comment:
This method will return either all input events hydrated or an error. An error can be returned only in the following scenarios:
1. context was canceled
2. request failed event after retries and client is configured to fail on error